linkerd/app/inbound/src/lib.rs

```diff
@@ -248,50 +249,41 @@ impl Config {
-                        .push_failfast(dispatch_timeout)
-                        .push_spawn_buffer_with_idle_timeout(buffer_capacity, cache_max_idle_age)
-                        .push(metrics.stack.layer(stack_labels("profile")))
-                        .box_http_response(),
-                ),
-            )
-            .into_make_service()
-            .spawn_buffer(buffer_capacity)
             .instrument(|_: &Target| debug_span!("profile"))
-            .check_make_service::<Target, http::Request<_>>();
+            .push_on_response(svc::layers().box_http_response())
+            .check_new_service::<Target, http::Request<http::boxed::Payload>>();
 
         let forward = target
-            .cache(
-                svc::layers().push_on_response(
-                    svc::layers()
-                        .push_failfast(dispatch_timeout)
-                        .push_spawn_buffer_with_idle_timeout(buffer_capacity, cache_max_idle_age)
-                        .push(metrics.stack.layer(stack_labels("forward")))
-                        .box_http_response(),
-                ),
-            )
-            .into_make_service()
-            .spawn_buffer(buffer_capacity)
             .instrument(|_: &Target| debug_span!("forward"))
-            .check_make_service::<Target, http::Request<http::boxed::Payload>>();
+            .check_new_service::<Target, http::Request<http::boxed::Payload>>();
 
         // Attempts to resolve the target as a service profile or, if that
         // fails, skips that stack to forward to the local endpoint.
         profile
-            .push_make_ready()
             .push_fallback(forward)
+            .check_new_service::<Target, http::Request<http::boxed::Payload>>()
             // If the traffic is targeted at the inbound port, send it through
-            // the loopback service (i.e. as a gateway). This is done before
-            // caching so that the loopback stack can determine whether it
-            // should cache or not.
+            // the loopback service (i.e. as a gateway).
             .push(admit::AdmitLayer::new(prevent_loop))
+            .check_new_service::<Target, http::Request<http::boxed::Payload>>()
             .push_fallback_on_error::<prevent_loop::LoopPrevented, _>(
                 svc::stack(loopback)
-                    .check_make_service::<Target, http::Request<_>>()
+                    .into_new_service()
+                    .check_new_service::<Target, http::Request<_>>()
                     .into_inner(),
             )
+            .check_new_service::<Target, http::Request<http::boxed::Payload>>()
+            .cache(
+                svc::layers().push_on_response(
+                    svc::layers()
+                        .push_failfast(dispatch_timeout)
+                        .push_spawn_buffer_with_idle_timeout(buffer_capacity, cache_max_idle_age)
+                        .push(metrics.stack.layer(stack_labels("logical")))
```

Comment:
nit: is this still per-logical?

Reply:
Yeah, the cache holds a service for each logical target (i.e. the apex service in a traffic split).